enhancements/sig-architecture/8-addon-framework/README.md

```diff
@@ -0,0 +1,338 @@
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		os.Exit(1)
+	}
+}
+
+func newCommand() *cobra.Command {
+	cmd := controllercmd.
+		NewControllerCommandConfig("addon-controller", version.Get(), runController).
+		NewCommand()
+	cmd.Use = "controller"
+	cmd.Short = "Start the addon controller"
+
+	return cmd
+}
+
+func runController(ctx context.Context, controllerContext *controllercmd.ControllerContext) error {
+	// Get bootstrap secret
+	kubeClient, err := kubernetes.NewForConfig(controllerContext.KubeConfig)
+	if err != nil {
+		return err
+	}
+	addonContext := &addonmanager.AddonManagerContext{
+		AddonName:      "helloworld",
+		AddonNamespace: "default",
+		KubeConfig:     controllerContext.KubeConfig,
+	}
+	agent := &helloWorldAgent{
+		kubeclient:        kubeClient,
+		hubhost:           "https://10.96.0.1",
+		controllerContext: controllerContext,
+		addonName:         addonContext.AddonName,
+	}
+	mgr := addonmanager.NewAddonManager(addonContext, agent).
+		WithRegistrationEnabled().
+      // A list of function to check whether the csr created by the addon agent should be approved by the manager
+      // Commonly, we need to check whether the requester is valid and whether CN/ORG field in the request is valid.
+      // If it is not set, only renewed csr is approved, and manual csr approve is needed for agent bootstrap
+		WithEnableCSRApproveFunc(
+			func(csr *certificates.CertificateSigningRequest) bool {
+				return true
+			},
+		)
+
+	mgr.Run(ctx)
+
+	<-ctx.Done()
+
+	return nil
+}
+
+type helloWorldAgent struct {
+	kubeclient        kubernetes.Interface
+	controllerContext *controllercmd.ControllerContext
+	hubhost           string
+	addonName         string
+}
+
+// This example only apply a configmap to the managedcluster
+func (h *helloWorldAgent) AgentManifests(cluster *clusterv1.ManagedCluster, config runtime.Object) ([]runtime.Object, error) {
+	return []runtime.Object{
+		&corev1.ConfigMap{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "ConfigMap",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "hello-world",
+				Namespace: "default",
+			},
+		},
+	}, nil
+}
+
+// No role/rolebinding is needed.
+func (h *helloWorldAgent) AgentHubRBAC(cluster *clusterv1.ManagedCluster, group string) (*rbacv1.Role, *rbacv1.RoleBinding) {
+	return nil, nil
+}
+
+func (h *helloWorldAgent) AgentBootstrapKubeConfig(cluster *clusterv1.ManagedCluster) ([]byte, error) {
+   // We create a service account as bootstrap credential in each cluster namespace.
+	return addonmanager.GenerateBootstrapKubeConfigBySA(
+		h.kubeclient, h.controllerContext.EventRecorder, h.hubhost, h.addonName, cluster.Name)
+}
+```
+
+### Test Plan
+
+E2E tests can use kind to create multiple clusters to test. An example should be created using addon framework for the purpose of testing
+
+- An addon agent can be successfully intalled on Cluster A
+- An addon agent have correct permission to connect to hub after registration
```

Comment:
how about self-join as a test path? it's better if there's a path for a meaningful e2e that doesn't require multiple clusters at the same time.

Reply:
added the  test case